README.md
# PDF Jeff
Feed Jeff a PDF and he'll turn it into images!

## Installation
1. Update the `config.php` file in `/utils` with your own settings.
2. Chuck Jeff into your Apache server and you're good to go! Jeff will create a folder called `data` in the root directory and store the PDFs there.

## Dependencies
Jeff requires the following dependencies:
* [ImageMagick](http://www.imagemagick.org/script/index.php)
* [PHP](http://php.net/)
* [Apache](http://httpd.apache.org/)
* [SleekDB](https://sleekdb.github.io)

Jeff uses SleekDB (a lightweight PHP database) to store conversion processes.

## Usage
There are two available requests, one for feeding Jeff a PDF and one for getting the images back and checking the processing status.

The intended use goes like this:
1. POST a PDF to Jeff.
2. Take the returned ID from the POST request and do a GET request to the same endpoint with the ID as a parameter.
3. The GET request will return the status of the conversion process. If the conversion is done, the images will be returned as well.
4. Repeat step 3 until the conversion is done (status will be set to "done" and images will go from null to an array of images than you can download).

### Feed Jeff a PDF
To feed Jeff a PDF, send a `POST` request to with the following `JSON` body:
```json
{
  "pdf": "https://example.com/example.pdf",
  "resolution": 300
}
```
The `resolution` parameter (DPI) is optional and defaults to 150.

Jeff will return a `JSON` object with the following structure:
```json
{
  "status": "processing",
  "message": "Images are being generated.",
  "id": "0cc62ed6e1ddd65cad328014f0802e138ffff39d"
}
```
### Get the images back (and check the status)
To get the images back, send a GET request to with the following query parameters:
* `id`: The ID of the conversion process

E.g. `https://pdfjeff.test/?id=0cc62ed6e1ddd65cad328014f0802e138ffff39d`

Jeff will initially return a `JSON` object with the following structure:
```json
{
  "status": "processing",
  "statusCode": 200,
  "message": "This file is currently being processed.",
  "images": null
}
```
If the status is still "processing", do another request with a delay of a few seconds. When Jeff has finished processing, the returned `JSON` object will look something like this:
```json
{
  "status": "done",
  "statusCode": 200,
  "message": "This file has finished processing. Images will automatically expire after 30 minutes.",
  "images": [
    "data/12a86ceed46ed2049fd9396edfaa1ec5b813f1e9/images/0.jpg",
    "data/12a86ceed46ed2049fd9396edfaa1ec5b813f1e9/images/1.jpg",
    "data/12a86ceed46ed2049fd9396edfaa1ec5b813f1e9/images/2.jpg",
    "data/12a86ceed46ed2049fd9396edfaa1ec5b813f1e9/images/3.jpg",
    "data/12a86ceed46ed2049fd9396edfaa1ec5b813f1e9/images/4.jpg",
    "data/12a86ceed46ed2049fd9396edfaa1ec5b813f1e9/images/5.jpg",
    "data/12a86ceed46ed2049fd9396edfaa1ec5b813f1e9/images/6.jpg",
    "data/12a86ceed46ed2049fd9396edfaa1ec5b813f1e9/images/7.jpg"
  ]
}
```
In case of a failure, Jeff will return a `JSON` object with the following structure:
```json
{
  "status": "error",
  "statusCode": 500,
  "message": "This file failed to process, please try again.",
  "images": null
}
```

## Issues
If you find anything that doesn't work as expected, feel free to open an issue using the GitHub issues tracker. Keep in mind that this is a hobby project and I'm not actively maintaining it. I don't intend for Jeff to be something he's not, so I (most likely) won't be adding any new features.

## License
Jeff is licensed under the MIT license. See the [LICENSE](LICENSE) file for more information.

![jeff](https://user-images.githubusercontent.com/1626458/202286502-1eb3abf0-73ff-46bb-b039-00e28a6b3317.gif)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2022-11-16

### Added

- This CHANGELOG file to hopefully serve as an evolving example of a standardized open source project CHANGELOG.
- A README file with a description of the project and documentation for installation and usage.
- A newborn Jeff.

utils/generate-images.php
<?php
// Database and configuration
require_once "config.php";
require_once dirname(__DIR__) . "/vendor/SleekDB/src/Store.php";
$databaseDirectory = dirname(__DIR__) . "/database";
$processes = new \SleekDB\Store("processes", $databaseDirectory);

try {
  // Arguments
  $id = $argv[1];
  $resolution = $argv[2];
  $processID = $argv[3];

  // Generate images with Imagick
  $image = new Imagick();
  $image->pingImage(DATA_PATH . $id . '/jeff.pdf');
  $pages = $image->getNumberImages();
  $images = array();
  for($i=0; $i < $pages; $i++) {
    $imagick = new Imagick();
    $imagick->setResolution($resolution,$resolution);
    $imagick->readImage(DATA_PATH . $id . '/jeff.pdf[' . $i . ']');
    $imagick->writeImage(DATA_PATH . $id . '/images/' . $i . '.jpg');
    $images[] = 'data/' . $id . '/images/' . $i . '.jpg';
  }

  // Update process in database with images and new status
  $updateddbprocess = $processes->updateById($processID, ["status" => "done", "timestamp" => strtotime("now"), "images" => $images]);
}
catch (ImagickException $e) {
  // Update process in database with failed status
  $updateddbprocess = $processes->updateById($processID, ["status" => "failed", "timestamp" => strtotime("now")]);
}


index.php
<?php
require_once "utils/config.php";
require_once "vendor/SleekDB/src/Store.php";

// Set headers
header('Content-Type: application/json; charset=utf-8');

// Database
$databaseDirectory = __DIR__ . "/database";
$processes = new \SleekDB\Store("processes", $databaseDirectory);

// Clear expired processes
$expiredProcesses = $processes->deleteBy(["timestamp", "<", (strtotime(EXPIRE_TIME))], 2);
for ($i = 0; $i < count($expiredProcesses); $i++) {
  // Delete directory and files for expired process
  exec('rm -rf data/' . $expiredProcesses[$i]["id"]);
}

if ($_SERVER['REQUEST_METHOD'] === 'POST') {
  // Decode JSON from request body
  try {
    $json = json_decode(file_get_contents('php://input'), false, 512, JSON_THROW_ON_ERROR);
    if (!isset($json->pdf)) {
      header("HTTP/1.1 400 Bad Request");
      echo json_encode(array('status' => 'Error', 'statusCode' => 400, 'message' => 'Mandatory parameters missing: pdf.'));
      return;
    }
  }
  catch (\JsonException $exception) {
    header('HTTP/1.1 400 Bad Request');
    echo json_encode(array('status' => 'Error', 'statusCode' => 400, 'message' => $exception->getMessage()));
    return;
  }

  // Parameters
  $resolution = isset($json->resolution) ? $json->resolution : 150; // Resolution in DPI
  $pdf = isset($json->pdf) ? $json->pdf : null; // PDF file
  $id = bin2hex(random_bytes(20)); // Unique ID

  // Create directories
  if(!is_dir('data')) {
    mkdir('data');
  }
  if(!is_dir('data/' . $id)) {
    mkdir('data/' . $id);
    mkdir('data/' . $id . '/images');
  }

  // Fetch PDF from URL with curl, save it to disk and check for errors
  $ch = curl_init($pdf);
  $fp = fopen('data/' . $id . '/jeff.pdf', 'wb');
  curl_setopt($ch, CURLOPT_FILE, $fp);
  curl_setopt($ch, CURLOPT_HEADER, 0);
  curl_exec($ch);
  curl_close($ch);
  fclose($fp);
  if(!file_exists('data/' . $id . '/jeff.pdf')) {
    header("HTTP/1.1 400 Bad Request");
    echo json_encode(array('status' => 'Error', 'message' => 'Could not fetch PDF.'));
    return;
  }

  // Save process to database
  $process = [
    "id" => $id,
    "resolution" => $resolution,
    "status" => "processing",
    "timestamp" => strtotime("now")
  ];
  $results = $processes->insert($process);

  // Generate images
  $info = shell_exec("php utils/generate-images.php '" . $id . "' ". $resolution . " " . $results['_id'] . " >/dev/null 2>&1 &");

  header("HTTP/1.1 200 OK");
  echo json_encode(array('status' => 'Processing', 'message' => 'Images are being generated.', 'id' => $id));

  return;
}
else if ($_SERVER['REQUEST_METHOD'] === 'GET') {
  // Fail with error if no ID is given
  $id = isset($_GET['id']) ? $_GET['id'] : null;
  if ($id == null) {
    header("HTTP/1.1 400 Bad Request");
    echo json_encode(array('status' => 'Error', 'statusCode' => 400, 'message' => 'Mandatory parameters missing: id.'));
    return;
  }

  // Find process in database
  $process = $processes->findOneBy(["id", "=", $id]);

  // Fail with error if process is not found
  if ($process == null) {
    header("HTTP/1.1 404 Not Found");
    echo json_encode(array('status' => 'Error', 'statusCode' => 404, 'message' => 'Process not found.'));
    return;
  }

  if($process['status'] === 'processing') {
    header("HTTP/1.1 200 OK");
    echo json_encode(array('status' => $process['status'], 'statusCode' => 200, 'message' => 'This file is currently being processed.', 'images' => isset($process['images']) ? $process['images'] : null));
  }
  elseif($process['status'] === 'done') {
    header("HTTP/1.1 200 OK");
    echo json_encode(array('status' => $process['status'], 'statusCode' => 200, 'message' => 'This file has finished processing. Images will automatically expire after 30 minutes.', 'images' => isset($process['images']) ? $process['images'] : null));
  }
  elseif($process['status'] === 'failed') {
    header("HTTP/1.1 500 Internal Server Error");
    echo json_encode(array('status' => $process['status'], 'statusCode' => 500, 'message' => 'This file failed to process, please try again.', 'images' => isset($process['images']) ? $process['images'] : null));
  }
}

?>